test/python/transpiler/test_clifford_passes.py

```diff
@@ -536,6 +537,122 @@ def test_do_not_merge_conditional_gates(self):
         # Make sure that the condition on the middle gate is not lost
         self.assertIsNotNone(qct.data[1].operation.condition)
 
+    def test_collect_with_cliffords(self):
+        """Make sure that collecting Clifford gates and replacing them by Clifford
+        works correctly when the gates include other cliffords."""
+
+        # Create a Clifford over 2 qubits
+        cliff_circuit = QuantumCircuit(2)
+        cliff_circuit.cx(0, 1)
+        cliff_circuit.h(0)
+        cliff = Clifford(cliff_circuit)
+
+        qc = QuantumCircuit(3)
+        qc.h(0)
+        qc.append(cliff, [1, 0])
+        qc.cx(1, 2)
+
+        # Collect clifford gates from the circuit (in this case all the gates must be collected).
+        qct = PassManager(CollectCliffords()).run(qc)
+        self.assertEqual(len(qct.data), 1)
+
+        # Make sure that the operator for the initial quantum circuit is equivalent to the
+        # operator for the collected clifford.
+        op1 = Operator(qc)
+        op2 = Operator(qct)
+        self.assertTrue(op1.equiv(op2))
```

Comment:
Will this test also work? 

```
cliff1 = Clifford(qc)
cliff2 = Clifford(qc2)
self.assertTrue(cliff1 == cliff2)
```

The same question applies to the other tests as well.

Reply:
Yes, it would, given the fix in `clifford_circuits.py` in this PR. Though please note that I have already added similar tests to `quantum_info/operators/symplectic/test_clifford`, i.e. that constructing a Clifford from quantum circuits (with other  Cliffords) produces the correct Clifford. Here in `transpiler/test_clifford_passes` I am more concerned with checking the correctness of the `CollectCliffords` transpiler pass (which under the hood does construct a Clifford from the given quantum circuit), so checking the equivalence of Operators seems slightly more appropriate.

Reply:
I agree with this test, but will `test_collect_with_linear_functions` , `test_collect_with_pauli_gates` and `test_collect_with_all_types` will work with this code?

Reply:
Thanks! So what was missing was a test in `quantum_info/operators/symplectic/test_clifford` that a Clifford can be constructed from a circuit that contains various clifford-like objects, that is regular clifford gates, linear functions, pauli gates, other cliffords, and other circuits with clifford-like objects. Added this test in [449cfcd](https://github.com/Qiskit/qiskit-terra/pull/9169/commits/449cfcda475ea2a88796345ee1d6966afd4cc023).